File: golang/executors/gorm_ex/cost.go
package gorm_ex

import (
	"bench-pg-go/model/domain"
	"gorm.io/gorm"
)

func accountCost(db *gorm.DB, op *domain.Cost, user domain.User) {
	tx := db.Begin()
	if tx.Error != nil {
		panic(tx.Error)
	}

	document := domain.NewCostDocument(
		op.PostingDate,
		op.DocNo,
		op.Material,
		op.Vendor,
		op.Amount,
		user,
	)
	postDocument(tx, document)

	res := tx.Commit()
	if res.Error != nil {
		panic(res.Error)
	}
}

File: rust/src/executors/diesel/cost.rs
use diesel::PgConnection;
use diesel::prelude::*;
use diesel::r2d2::{ConnectionManager, PooledConnection};
use diesel::result::Error;

use super::post_document::post_document;
use crate::model::domain::{Cost, Document, User};

pub async fn account_cost(
    conn: &mut PooledConnection<ConnectionManager<PgConnection>>,
    op: &Cost,
    user: User,
) {
    conn.transaction(|tx| {
        let document = Document::new_cost_document(
            op.posting_date,
            op.doc_no.clone(),
            op.material,
            op.vendor,
            op.amount,
            user,
        );

        post_document(tx, document).expect("Cannot post cost document");

        Ok::<(), Error>(())
    })
    .expect("Cannot commit transaction for account material cost");
}
